gammapy/detect/cwt.py

```diff
@@ -57,200 +62,552 @@ def difference_of_gauss_kernel(radius, scale_step, n_sigmas=8):
 
-    - computes significance ``sig = transform / error``
-    - support : 3D cube exclusion mask
-      - filled as ``sig > nsigma``
+    def _is_converged(self, data):
+        """
+        Check if the algorithm has converged on current iteration.
 
-    The ``inverse_transform`` step does the following:
+        TODO: document metric used, but not super important.
 
-    - model : 2D image, init to 0
-        - Fill ``res = np.sum(self.support * self.transform, axis=0)``
-        - Fill ``self.model += res * (res > 0)``
-        - What is this??
+        Parameters
+        ----------
+        data : `~gammapy.detect.CWTData`
+            Images after iteration.
 
+        Returns
+        -------
+        answer : boolean
+            Answer if CWT has converged.
+        """
+        answer = False
+        residual = data.counts - (data.model + data.approx)
+        variance = residual.var()
+        if self.previous_variance is not None:
+            variance_ratio = abs((self.previous_variance - variance) / self.previous_variance)
+            answer = variance_ratio < self.tol
+            log.info('Variance ratio: {:.7f}'.format(variance_ratio))
+        self.previous_variance = variance
+        log.info('')
+        return answer
+
+    def analyze(self, data):
+        """
+        Run iterative filter peak algorithm. The algorithm modifies the original data.
 
-    - max_scale_image : 2D image, value = index of scale where emission is most significant
+        Parameters
+        ----------
+        data : `~gammapy.detect.CWTData`
+            Input images.
+        """
+        if self.history is not None:
+            self.history = [copy.deepcopy(data)]
+
+        for n_iter in range(self.max_iter):
+            log.info('Start iteration {0}'.format(n_iter + 1))
+            self._execute_iteration(data=data)
+            if self.history is not None:
+                self.history.append(copy.deepcopy(data))
+            if self._is_converged(data=data):
+                break
+
+        if n_iter != self.max_iter:
+            log.info('Convergence reached at iteration {0}'.format(n_iter + 1))
+        else:
+            log.info('Convergence not formally reached at iteration {0}'.format(n_iter + 1))
+
+    def get_history(self):
+        """
+        Return cwt data from all the iterations.
+
+        Returns
+        -------
+        data : list
+            List of '~gamapy.detect.CWTData' objects.
+        """
+        if self.history is None:
+            raise AttributeError('This CWT object does not support saving cwt data images.')
+        else:
+            return copy.deepcopy(self.history)
+
+
+class CWTKernels(object):
+    """
+    Conduct arrays of kernels and scales for CWT algorithm.
+
+    Parameters
+    ----------
+    n_scale : int
+        Number of scales.
+    min_scale : double
+        First scale used.
+    step_scale : double
+        Base scaling factor.
+    old : boolean (default True)
```

Comment:
Did you investigate the difference between old and new way to compute the kernels?
If no -- should I do it or explain how to?

Once that's done and results understood this option "old" in CWTKernels should be removed.